kafka-store/src/test/java/io/confluent/kafka/schemaregistry/ClusterTestHarness.java

```diff
@@ -0,0 +1,178 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.confluent.kafka.schemaregistry;
+
+import kafka.utils.CoreUtils;
+import org.I0Itec.zkclient.ZkClient;
+import org.apache.kafka.common.protocol.SecurityProtocol;
+import org.junit.After;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.util.List;
+import java.util.Properties;
+import java.util.Vector;
+
+import io.confluent.kafka.schemaregistry.storage.KafkaStoreConfig;
+import kafka.server.KafkaConfig;
+import kafka.server.KafkaServer;
+import kafka.utils.SystemTime$;
+import kafka.utils.TestUtils;
+import kafka.utils.ZkUtils;
+import kafka.zk.EmbeddedZookeeper;
+import scala.Option;
+import scala.Option$;
+import scala.collection.JavaConversions;
+
+/**
+ * Test harness to run against a real, local Kafka cluster and REST proxy. This is essentially
+ * Kafka's ZookeeperTestHarness and KafkaServerTestHarness traits combined and ported to Java with
+ * the addition of the REST proxy. Defaults to a 1-ZK, 3-broker, 1 REST proxy cluster.
+ */
+public abstract class ClusterTestHarness {
+
+  public static final int DEFAULT_NUM_BROKERS = 1;
+  public static final String KAFKASTORE_TOPIC = KafkaStoreConfig.KAFKASTORE_CONFIG_PREFIX + KafkaStoreConfig.DEFAULT_KAFKASTORE_TOPIC;
+  protected static final Option<Properties> EMPTY_SASL_PROPERTIES = Option$.MODULE$.<Properties>empty();
+
+  /**
+   * Choose a number of random available ports
+   */
+  public static int[] choosePorts(int count) {
+    try {
+      ServerSocket[] sockets = new ServerSocket[count];
+      int[] ports = new int[count];
+      for (int i = 0; i < count; i++) {
+        sockets[i] = new ServerSocket(0, 0, InetAddress.getByName("0.0.0.0"));
+        ports[i] = sockets[i].getLocalPort();
+      }
+      for (int i = 0; i < count; i++)
+        sockets[i].close();
+      return ports;
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Choose an available port
+   */
+  public static int choosePort() {
+    return choosePorts(1)[0];
+  }
+
+  private int numBrokers;
+
+  // ZK Config
+  protected EmbeddedZookeeper zookeeper;
+  protected String zkConnect;
+  protected ZkClient zkClient;
+  protected ZkUtils zkUtils;
+  protected int zkConnectionTimeout = 30000; // a larger connection timeout is required for SASL tests
+                                             // because SASL connections tend to take longer.
+  protected int zkSessionTimeout = 6000;
+
+  // Kafka Config
+  protected List<KafkaConfig> configs = null;
+  protected List<KafkaServer> servers = null;
+  protected String brokerList = null;
+
+  //protected RestApp restApp = null;
```

Comment:
Can we avoid a lot of code duplication here with a bit of inheritance and overrides? It seems like very little changed aside from removing a few lines related to instantiating and starting the rest app?